examples/step-63/step-63.cc

```diff
@@ -19,16 +19,1089 @@
+                             Smoother::AdditionalData(fe.degree == 1 ? 0.6667 :
+                                                                       0.47));
+        smoother->set_steps(4);
+        mg_smoother = std::move(smoother);
+      }
+    else if (settings.smoother_type == "block SOR")
+      {
+        using Smoother =
+          RelaxationBlockSOR<SparseMatrix<double>, double, Vector<double>>;
+
+        smoother_data.resize(0, triangulation.n_levels() - 1);
+
+        for (unsigned int level = 0; level < triangulation.n_levels(); ++level)
+          {
+            DoFTools::make_cell_patches(smoother_data[level].block_list,
+                                        dof_handler,
+                                        level);
+
+            smoother_data[level].relaxation = 1.0;
+            smoother_data[level].inversion = PreconditionBlockBase<double>::svd;
+
+            std::vector<unsigned int> ordered_indices;
+            switch (settings.dof_renumbering)
+              {
+                case Settings::DoFRenumberingStrategy::downstream:
+                  ordered_indices =
+                    create_downstream_cell_ordering(dof_handler,
+                                                    advection_direction,
+                                                    level);
+                  break;
+
+                case Settings::DoFRenumberingStrategy::upstream:
+                  ordered_indices =
+                    create_downstream_cell_ordering(dof_handler,
+                                                    -1.0 * advection_direction,
+                                                    level);
+                  break;
+
+                case Settings::DoFRenumberingStrategy::random:
+                  ordered_indices =
+                    create_random_cell_ordering(dof_handler, level);
+                  break;
+
+                case Settings::DoFRenumberingStrategy::none: // Do nothing
+                  break;
+
+                default:
+                  AssertThrow(false, ExcNotImplemented());
+                  break;
+              }
+
+            smoother_data[level].order =
+              std::vector<std::vector<unsigned int>>(1, ordered_indices);
+          }
+
+        auto smoother =
+          std_cxx14::make_unique<MGSmootherPrecondition<SparseMatrix<double>,
+                                                        Smoother,
+                                                        Vector<double>>>();
+        smoother->initialize(mg_matrices, smoother_data);
+        smoother->set_steps(1);
+        mg_smoother = std::move(smoother);
+      }
+    else if (settings.smoother_type == "block Jacobi")
+      {
+        using Smoother =
+          RelaxationBlockJacobi<SparseMatrix<double>, double, Vector<double>>;
+
+        smoother_data.resize(0, triangulation.n_levels() - 1);
+
+        for (unsigned int level = 0; level < triangulation.n_levels(); ++level)
```

Comment:
this is copy&pasted from above, right? Can we extract this into a method?

Reply:
see my comment below.